### PR TITLE
Feature/sync GitHub button

### LINKED ISSUE #234: Create button for manual sync of github data
User should be able to obtain realtime data any time s.he wishes to apart from the everyday export activity.

Provide UI (button) 
Also keep a text log to include export paramters and status ( required to detect errors if export is unsuccessful)

# Definition of done
Choose at least one of these tasks:

- [x] Button exists for manual syncing of Github data for a given data source
  - Issues
  - Commits
- [x] Button exists for manual syncing of Github data for a given service
- [x] When clicking "sync" button, all relevant Github data is fetched

### PR DESCRIPTION
**Pull request by Alapan**
Closes #234 

# Proposed changes
(List any significant changes or additions here)
- Added a 'Sync' button for each source under a service.
- When the button is pressed, all commit and issues for that source is fetched.
- The local collections 'GithubIssues' and 'GithubCommits' are checked to see if the fetched documents exist. If not, the collections are updated.
- From the client side, the previous point is executed by `githubSyncCommits` and `githubSyncRepoIssues` methods. 

@Digipalvelutehdas/dipor-team please review.

### CODE REVIEW

File: services/client/single/sources/sources.html [hunk 1]
```diff
@@ -23,6 +23,9 @@
         <th>
           Commits Activity
         </th>
+        <th>
+          Sync local with remote
```
Comment (brylie) [inline services/client/single/sources/sources.html]: This string needs to be translated, so add i18n token here. Also, add a shorter version of this text to the en.i18n.json file.

File: services/client/single/sources/syncGithubButton/sync-button.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<template name="syncGithubButton">
+  <button type="button" class="btn btn-primary" id="sync-github-data">Sync</button>
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.html]: Use `class` instead of `id`. Id is used when there will be only one of a given dom element in the document. Since there are multiple sync buttons on the page, i.e. one per table row, the `class` should be used.

# Reference
* CSS Tricks - [The Difference Between Id and Class](https://css-tricks.com/the-difference-between-id-and-class/)

Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.html]: Also, consider making the button smaller, so that it is not so visually prominent.

# Resource
- Check out the [Bootstrap Button Builder](http://bootsnipp.com/buttons)

File: services/client/single/sources/syncGithubButton/sync-button.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
+Template.syncGithubButton.events({
+
+  'click #sync-github-data': (event, template) => {
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.js]: Remember to change the `#` to a `.` when switching from ID to class.

File: services/client/single/sources/syncGithubButton/sync-button.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    // get references to username and repository
+    const username = template.data.user;
+    const repository = template.data.repo;
+
+    // call methods for syncing issue and commit data with local GithubIssues and GithubCommits collections
+    Meteor.call('githubSyncCommits', username, repository, (error) => {
+      // display message if syncing is not successful
+      if (error) {
+        alert(error);
+      }
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.js]: How can we let the user know if/when the sync was successful?

**Comment by brylie** [comment]
@Alapan status update?

**Comment by Alapan** [comment]
I am writing the code now for showing a successful update to the user. But I am not yet sure which glyphicon to use for making the button seem smaller.

**Comment by Alapan** [comment]
@brylie I made the changes you suggested. Please review.

**Comment by brylie** [comment]
Looks good so far. However, I am getting an error in the server console. I opened issue #267 regarding the error.

**Comment by brylie** [comment]
This is pretty much ready to merge. One thing more would be to use the [s-alert](https://github.com/juliancwirko/meteor-s-alert#success) success message, rather than the browser `alert()` method. S-alert is a bit more user-friendly.

File: services/client/single/sources/syncGithubButton/sync-button.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    const username = template.data.user;
+    const repository = template.data.repo;
+
+    // call methods for syncing issue and commit data with local GithubIssues and GithubCommits collections
+    Meteor.call('githubSyncRepoCommits', username, repository, (error, commitActivity) => {
+      // display message if syncing is not successful
+      if (error) {
+        alert(error);
+      } else {
+        alert("Sync commits successful!");
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.js]: Try using [s-alert successs](https://github.com/juliancwirko/meteor-s-alert#success) here, instead of `alert()`.

File: services/client/single/sources/syncGithubButton/sync-button.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      }
+    });
+
+    Meteor.call('githubSyncRepoIssues', username, repository, (error, issueActivity) => {
+      if (error) {
+        alert(error);
+      } else {
+        alert("Sync issues successful!");
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.js]: Try using [s-alert successs](https://github.com/juliancwirko/meteor-s-alert#success) here, instead of `alert()`.

**Comment by Alapan** [comment]
@brylie The change was made. Please review.

**Comment by Nazarah** [comment]
Add an indication that the sync request is in progress.

**Comment by brylie** [comment]
@Alapan, we have some example code for toggling the Bootstrap button state when clicked, so that the user gets feedback that their request is in progress.

**Comment by Alapan** [comment]
@brylie I added the `sacha:spin` package for showing a spinner, and have written some code for showing it, but it shows `instance.serverCallStarted is undefined` when using `Template.syncGithubButton.onCreated(() => {` iin place of `Template.syncGithubButton.onCreated((function() {`, and with the latter syntax, does not show the error, but also does not show the spinner. Should I use something else instead?

**Comment by brylie** [comment]
One thing is that `instance.serverCallStarted` should be a reactive variable. Otherwise, code depending on it will not re-run.

```
Template.syncGithubButton.onCreated(function () {
  // Get reference to template instance
  const instance = this;

  // Create reactive variable to track sync status
  instance.serverCallRunning = new ReactiveVar();
});

**Comment by Alapan** [comment]
I had made all those changes you suggested earlier, and removed the ES6 syntax for functions used earlier, but the spinner is not showing on my machine.

**Comment by brylie** [comment]
Alright, have you tested the spinner in another template or without any conditional logic? E.g. what happens if you simply include the spinner in the front page template?

**Comment by brylie** [comment]
@elnzv is there a simpler approach to this task?

File: .meteor/packages~ [hunk 7]
```diff
@@ -0,0 +1,56 @@
+# Meteor packages used by this project, one per line.
```
Comment (brylie) [inline .meteor/packages~]: This file should not be included in the PR.

File: .meteor/versions~ [hunk 8]
```diff
@@ -0,0 +1,144 @@
+accounts-base@1.2.7
```
Comment (brylie) [inline .meteor/versions~]: This file should not be included in the PR.

File: core/lib/i18n/en.i18n.json~ [hunk 9]
```diff
@@ -0,0 +1,89 @@
+{
```
Comment (brylie) [inline core/lib/i18n/en.i18n.json~]: This file should not be included in the PR.

**Comment by brylie** [comment]
@Alapan be careful with each of your commits that you are aware of what is being added and that it should be included in the commit. There are several temporary files that are included in this PR and were possibly inadvertently committed.

**Comment by Alapan** [comment]
I really don't know how so many temporary files got committed. I had committed the files I had changed individually.

**Comment by elnzv** [comment]
@Alapan, I will suggest not to use spinner or any other loading stage packages for now.
There is a good and easy way how to show to the user that there is a synchronization going on.
Just use jQuery to change the button, example:

```
      $('.sync-github-data').addClass('disabled').attr('disabled', true);
      $('.sync-github-data').text('Syncing...');

**Comment by Alapan** [comment]
@elnzv Thanks a lot for the suggestion! It works. 
@brylie I removed the spinner logic, and now am only toggling the button text and state.

**Comment by brylie** [comment]
Excellent! I like the simplicity of that solution.

File: services/client/single/sources/syncGithubButton/sync-button.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    // disable button and change text when pressed, to indicate call in progress
+    $(event.target).prop("disabled", true).text("Syncing...");
+
+    // call methods for syncing issue and commit data with local GithubIssues and GithubCommits collections
+    Meteor.call('githubSyncRepoCommits', username, repository, (error, commitActivity) => {
+      // display message if syncing is not successful
+      if (error) {
+        alert(error);
+      } else {
+        sAlert.success("Syncing commits successful!");
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.js]: We need to also internationalize these texts. Thankfully, the TAP i18n library has JS methods for retrieving a given key:

```js
// notice the double underscore
const syncCommitsSuccessMessage = TAPi18n.__("syncGithubButton-syncCommitsSuccessMessage");

sAlert.success(syncCommitsSuccessMessage);
```

File: services/client/single/sources/syncGithubButton/sync-button.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
+Template.syncGithubButton.events({
+
+  'click .sync-github-data': function (event, template) {
+
+    // get references to username and repository
+    const username = template.data.user;
+    const repository = template.data.repo;
+
+    // disable button and change text when pressed, to indicate call in progress
+    $(event.target).prop("disabled", true).text("Syncing...");
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.js]: Make sure to internationalize this string "Syncing...".

```js
// notice the double underscore
const syncMessage = TAPi18n.__("syncGithubButton-syncMessage");

sAlert.success(syncMessage);
```

File: services/client/single/sources/syncGithubButton/sync-button.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+      } else {
+        sAlert.success("Syncing commits successful!");
+      }
+    });
+
+    Meteor.call('githubSyncRepoIssues', username, repository, function (error, issueActivity){
+      if (error) {
+        alert(error);
+      } else {
+        sAlert.success("Syncing issues successful!");
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.js]:
```js
// notice the double underscore
const syncIssuesSuccessMessage = TAPi18n.__("syncGithubButton-syncIssuesSuccessMessage");

sAlert.success(syncIssuesSuccessMessage);
```

File: services/client/single/sources/syncGithubButton/sync-button.html [hunk 13]
```diff
@@ -0,0 +1,3 @@
+<template name="syncGithubButton">
+  <button type="button" class="btn btn-primary btn-sm sync-github-data">Sync</button>
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.html]: Add i18n token here.

```
{{_ "syncGithubButton-syncButton" }}
```

Note: do not re-use translation strings.

File: services/client/single/sources/syncGithubButton/sync-button.js [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+      }
+      // calls completed, now enable button and change button text back to 'Sync'
+      $(event.target).prop("disabled", false).text("Sync");
```
Comment (brylie) [inline services/client/single/sources/syncGithubButton/sync-button.js]: Add i18n string here.

**Comment by brylie** [comment]
OK, this task is really close. The last part is to add i18n tokens/strings for all user facing text. You don't have to translate error messages, just buttons, status indicators, etc.

**Comment by Alapan** [comment]
The changes were made. Please review.
